src/typechecker.erl

```diff
@@ -953,45 +957,65 @@ expect_list_type(Union = {type, _, union, UnionTys}, N) ->
         _ ->
             {elem_tys, Tys, Cs}
     end;
-expect_list_type({var, _, Var}, _) ->
+expect_list_type({var, _, Var}, _, _) ->
     TyVar = new_type_var(),
     {elem_ty
     ,{var, erl_anno:new(0), TyVar}
     ,constraints:add_var(TyVar,
       constraints:upper(Var, {type, erl_anno:new(0), list, [TyVar]}))
     };
-expect_list_type(Ty, _) ->
+expect_list_type(Ty, _, _) ->
     {type_error, Ty}.
 
 
-expect_list_union([Ty|Tys], AccTy, AccCs, Any, N) ->
-    case expect_list_type(Ty, N) of
+expect_list_union([Ty|Tys], AccTy, AccCs, Any, N, TEnv) ->
+    case expect_list_type(normalize(Ty, TEnv), N, TEnv) of
         {type_error, _} ->
-            expect_list_union(Tys, AccTy, AccCs, Any, N);
+            expect_list_union(Tys, AccTy, AccCs, Any, N, TEnv);
         any ->
             expect_list_union(Tys
                              ,[type(any) | AccTy]
                              ,AccCs
                              ,any
-                             ,N);
+                             ,N
+                             ,TEnv);
         {elem_ty, NTy, Cs} ->
             expect_list_union(Tys
                              ,[NTy | AccTy]
                              ,constraints:combine(Cs, AccCs)
                              ,Any
-                             ,N);
+                             ,N
+                             ,TEnv);
         {elem_tys, NTys, Cs} ->
             expect_list_union(Tys
                              ,NTys ++ AccTy
                              ,constraints:combine(Cs, AccCs)
                              ,Any
-                             ,N)
+                             ,N
+                             ,TEnv)
     end;
-expect_list_union([], AccTy, AccCs, any, _N) ->
+expect_list_union([], AccTy, AccCs, any, _N, _TEnv) ->
     {[type(any) | AccTy], AccCs};
-expect_list_union([], AccTy, AccCs, _NoAny, _N) ->
+expect_list_union([], AccTy, AccCs, _NoAny, _N, _TEnv) ->
     {AccTy, AccCs}.
 
+infer_literal_string("") ->
+    type(nil);
+infer_literal_string(Str) ->
+    SortedChars = lists:usort(Str),
+    case lists:all(fun(C) -> C =< 255 end, SortedChars) of
+        _ when length(SortedChars) =< 10 ->
+            %% heuristics: if there are not more than 10 different charactes
+            %% list them explicitely as singleton types
+            CharTypes = [{char, erl_anno:new(0), C} || C <- SortedChars],
+            type(nonempty_list, [type(union, CharTypes)]);
```

Comment:
Normalize the union to so ranges become ranges?

Reply:
good suggestion